runtime/src/accounts_db.rs

```diff
@@ -3650,24 +3661,54 @@ impl AccountsDB {
         let result: Vec<_> = (0..chunks)
             .into_par_iter()
             .map(|i| {
-                let start_index = i * fanout;
-                let end_index = std::cmp::min(start_index + fanout, total_hashes);
+                let start_index = i * num_hashes_per_chunk;
+                let end_index = std::cmp::min(start_index + num_hashes_per_chunk, total_hashes);
 
                 let mut hasher = Hasher::default();
                 let mut data_index = start_index;
                 let mut data = data;
                 let mut data_len = data_len;
 
-                for i in start_index..end_index {
-                    if data_index >= data_len {
-                        // fetch next slice
-                        data = get_hashes(i);
-                        data_len = data.len();
-                        data_index = 0;
+                if !three_level {
+                    // 1 group of fanout
+                    // The result of this loop is a single hash value from fanout input hashes.
+                    for i in start_index..end_index {
+                        if data_index >= data_len {
+                            // fetch next slice
+                            data = get_hashes(i);
+                            data_len = data.len();
+                            data_index = 0;
+                        }
+                        hasher.hash(data[data_index].as_ref());
+                        data_index += 1;
+                    }
+                } else {
+                    // hash 3 levels of fanout simultaneously.
+                    // The result of this loop is a single hash value from fanout^3 input hashes.
+                    let mut i = start_index;
+                    while i < end_index {
+                        let mut hasher_j = Hasher::default();
+                        for _j in 0..fanout {
+                            let mut hasher_k = Hasher::default();
```

Comment:
ok, we have tests that cover all the edge-cases here?

Reply:
The case of having a partial last block is specifically tested with and without the 3 level optimization and compared.
I just made the test below more thorough to saturate the entire partial chunk solution space for a fanout of 3.
https://github.com/solana-labs/solana/blob/4ab1678b5b6c7859ec05ced8bc796588aec1a7a3/runtime/src/accounts_db.rs#L6544
In general the several merkle variations are tried with various calls and the results are compared with each other.
Note that the merkle code paths are about to be simplified in an upcoming PR as it will now be efficient for both uses - whether inputs are [Hash] or [(Pubkey, Hash)] where either of these types of input could be represented in 1-d to 3-d input vectors.

Reply:
OK. thanks